ospd_openvas/notus.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2014-2021 Greenbone Networks GmbH
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+from pathlib import Path
+from typing import Dict, Optional
+import json
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class Notus:
+    # caches the oids as well as the filepath to load the meta_data
+    # probably better in redis
+    cache: Dict[str, str]
+    path: Path
+
+    def __init__(self, path: str):
+        self.path = Path(path)
+        self.cache = {}
+
+    def reload_cache(self):
+        new_cache = {}
+        for f in self.path.glob('*.notus'):
+            filename = f.as_posix()
+            data = json.loads(f.read_bytes())
+            advisories = data.get("advisories", [])
+            for advisory in advisories:
+                new_cache[advisory["oid"]] = filename
+        self.cache = new_cache
+
+    def get_filenames_and_oids(self):
+        if not self.cache:
+            self.reload_cache()
+        for key, value in self.cache.items():
+            yield (Path(value).name, key)
+
+    def find_advisory(self, path: Path, oid: str):
+        advisories = json.loads(path.read_bytes())
+        for advisory in advisories.get("advisories", []):
+            if advisory.get("oid") == oid:
+                return advisory
```

Comment:
When getting all vts or doing the integrity check the program always loads the whole file (~11mb for euleros) for each vt. This causes the the integrity check and get_vts OSP command to take about 20 min. Alternatively save all nvt metadata within the nvt cache so reading the file has to be only done once at startup and each time the feed gets updated.